oshmem/mca/spml/ucx/spml_ucx.c

```diff
@@ -126,61 +127,120 @@ int mca_spml_ucx_del_procs(ompi_proc_t** procs, size_t nprocs)
     /* No need to barrier here - barrier is called in _shmem_finalize */
     free(del_procs);
-    free(mca_spml_ucx.remote_addrs_tbl);
+    if (mca_spml_ucx.remote_addrs_tbl) {
+        for (w = 0; w < ucp_workers; w++) {
+            if (mca_spml_ucx.remote_addrs_tbl[w]) {
+                for (n = 0; n < nprocs; n++) {
+                    if (mca_spml_ucx.remote_addrs_tbl[w][n]) {
+                        free(mca_spml_ucx.remote_addrs_tbl[w][n]);
+                    }
+                }
+                free(mca_spml_ucx.remote_addrs_tbl[w]);
+            }
+        }
+        free(mca_spml_ucx.remote_addrs_tbl);
+    }
+
     free(mca_spml_ucx_ctx_default.ucp_peers);
 
     mca_spml_ucx_ctx_default.ucp_peers = NULL;
 
     return ret;
 }
 
+static void dump_address(char *s, int pe, char *addr, size_t len)
+{
+#ifdef SPML_UCX_DEBUG
+    int my_rank = oshmem_my_proc_id();
+    unsigned  i;
+
+    printf("%s e=%d dest_pe=%d addr=%p len=%d\n", s, my_rank, pe, addr, len);
+    for (i = 0; i < len; i++) {
+        printf("%02X ", (unsigned)0xFF&addr[i]);
+    }
+    printf("\n");
+#endif
+}
+
 /* TODO: move func into common place, use it with rkey exchng too */
 static int oshmem_shmem_xchng(
-        void *local_data, int local_size, int nprocs,
-        void **rdata_p, int **roffsets_p, int **rsizes_p)
+        void **local_data, unsigned int *local_size, int nprocs, int ucp_workers,
+        void **rdata_p, unsigned int **roffsets_p, unsigned int **rsizes_p)
 {
-    int *rcv_sizes   = NULL;
-    int *rcv_offsets = NULL; 
-    void *rcv_buf    = NULL;
+    unsigned int *rcv_sizes   = NULL;
+    int *_rcv_sizes = NULL;
+    unsigned int *rcv_offsets = NULL;
+    int *_rcv_offsets = NULL; 
+    void *rcv_buf       = NULL;
     int rc;
-    int i;
+    int i,j,k;
 
-    /* do llgatherv */
-    rcv_offsets = malloc(nprocs * sizeof(*rcv_offsets));
+    /* do allgatherv */
+    rcv_offsets = calloc(ucp_workers * nprocs, sizeof(*rcv_offsets));
     if (NULL == rcv_offsets) {
         goto err;
     }
 
     /* todo: move into separate function. do allgatherv */
-    rcv_sizes = malloc(nprocs * sizeof(*rcv_sizes));
+    rcv_sizes = calloc(ucp_workers * nprocs, sizeof(*rcv_sizes));
     if (NULL == rcv_sizes) {
         goto err;
     }
-    
-    rc = oshmem_shmem_allgather(&local_size, rcv_sizes, sizeof(int));
+   
+    rc = oshmem_shmem_allgather(local_size, rcv_sizes, ucp_workers * sizeof(*rcv_sizes));
     if (MPI_SUCCESS != rc) {
         goto err;
     }
 
     /* calculate displacements */
     rcv_offsets[0] = 0;
-    for (i = 1; i < nprocs; i++) {
+    for (i = 1; i < ucp_workers * nprocs; i++) {
         rcv_offsets[i] = rcv_offsets[i - 1] + rcv_sizes[i - 1];
     }
 
-    rcv_buf = malloc(rcv_offsets[nprocs - 1] + rcv_sizes[nprocs - 1]);
+    rcv_buf = malloc(rcv_offsets[(ucp_workers * nprocs) - 1]
+                        + rcv_sizes[(ucp_workers * nprocs) - 1]);
```

Comment:
You switched to `calloc` above, why keep it `malloc`?

Reply:
I kept it mostly for readability, above we are dealing with elements of sizeof() so switched to highlight that (elements, sizeof(element). Here we're dealing with an absolute size so kept it as malloc, I thought that calloc (1, size) would look off. Rather than reverting to malloc; however, and for consistency, I'll change this line to calloc since calloc is prevalent.